### PR TITLE
glfw: 3.3.10 -> 3.4

### PR DESCRIPTION
**Pull request by getchoo**
## Description of changes

Release notes: https://github.com/glfw/glfw/releases/tag/3.4
Diff: https://github.com/glfw/GLFW/compare/3.3.10...3.4

Notable changes include better Wayland support, and the [deprecation](https://github.com/glfw/glfw/blob/3.4/CMakeLists.txt#L32) of `GLFW_USE_WAYLAND`.

I would also suggest following upstream here and enabling both X11 and Wayland on supported platforms (which is done by default), removing the need for the `glfw-wayland` package; this hasn't been implemented for now though as I wanted to get some feedback.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by getchoo** [comment]
it looks like this can also allow for the removal of `glfw-wayland-minecraft`, though we should probably link libdecor here as well
image

**Comment by NovaViper** [comment]
I just tried the PR with Minecraft 1.20.4 but the flickering is still there unfortunately in the XWayland mode but it's not nearly as bad as it was before! I'm wondering how I could make it run in Wayland instead of XWayland 🤔 
```
❯ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.8.0, NixOS, 24.05 (Uakari), 24.05.20240315.2bcf18c`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - nixpkgs: `/nix/store/y2k721a8jamdqxqbilyk6ds6374ljcgr-source`
```

**Comment by getchoo** [comment]
> I'm wondering how I could make it run in Wayland instead of XWayland

try overriding prism with `glfw-wayland` (i.e., `prismlauncher.override { glfw = pkgs.glfw-wayland; }`) and enabling native glfw in your instance's settings

we could just remove this package and give `glfw` wayland support, though

**Comment by NovaViper** [comment]
@getchoo Hey sorry for the delay! I can get Minecraft working using `glfw-wayland-minecraft` (with `prismlauncher-qt5 = prev.prismlauncher-qt5.override { withWaylandGLFW = true; };`) makes the game run in Wayland and it doesn't crash (but the "Wayland platform does not support setting cursor position" error appears in the logs whenever I open any of the GUI elements like the inventory). Looks like the minecraft version of `glfw-wayland` that's currently in nixpkgs is actually old and [hasn't had the patch needed to make the cursor position work properly](https://github.com/Admicos/minecraft-wayland/pull/56).

However, when I try to use this PR and override glfw either like this
```
    prismlauncher-qt5 = prev.prismlauncher-qt5.override {
      glfw = inputs.nixpkgs-glfw.legacyPackages.${prev.system}.glfw-wayland;
    };
````
or this
```
     glfw = inputs.nixpkgs-glfw.legacyPackages.${prev.system}.glfw.override {
       waylandSupport = true;
     };
    # use prismlauncher-qt without any additional override
```

The game now crashes with this error
```
[13:09:23] [Render thread/WARN]: Failed to create window: 
net.minecraft.class_1041$class_4716: GLFW error 65548: Wayland: The platform does not support setting the window icon
	at net.minecraft.class_1041.method_4501(class_1041.java:196) ~[client-intermediary.jar:?]
	at org.lwjgl.glfw.GLFWErrorCallbackI.callback(GLFWErrorCallbackI.java:43) ~[lwjgl-glfw-3.3.1.jar:?]
	at org.lwjgl.system.JNI.invokePPV(Native Method) ~[lwjgl-3.3.1.jar:?]
	at org.lwjgl.glfw.GLFW.nglfwSetWindowIcon(GLFW.java:2336) ~[lwjgl-glfw-3.3.1.jar:?]
	at org.lwjgl.glfw.GLFW.glfwSetWindowIcon(GLFW.java:2368) ~[lwjgl-glfw-3.3.1.jar:?]
	at net.minecraft.class_1041.method_4491(class_1041.java:175) ~[client-intermediary.jar:?]
	at net.minecraft.class_310.<init>(class_310.java:492) ~[client-intermediary.jar:?]
	at net.minecraft.client.main.Main.main(Main.java:211) ~[minecraft-1.20.1-client.jar:?]
	at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:470) ~[fabric-loader-0.15.7.jar:?]
	at net.fabricmc.loader.impl.launch.knot.Knot.launch(Knot.java:74) ~[fabric-loader-0.15.7.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClient.main(KnotClient.java:23) ~[fabric-loader-0.15.7.jar:?]
	at org.prismlauncher.launcher.impl.StandardLauncher.launch(StandardLauncher.java:87) ~[NewLaunch.jar:?]
	at org.prismlauncher.EntryPoint.listen(EntryPoint.java:129) ~[NewLaunch.jar:?]
	at org.prismlauncher.EntryPoint.main(EntryPoint.java:70) ~[NewLaunch.jar:?]
Process exited with code 0.
```

So there's definitely something in the `glfw-wayland-minecraft` package that includes changes that makes the game launch in Wayland properly, I'm just unsure of what exactly though 🤷🏾‍♀️

**Comment by NovaViper** [comment]
Also to add, when I add the PR glfw like this to PrismLauncher:
```
    glfw = inputs.nixpkgs-glfw.legacyPackages.${prev.system}.glfw.override {
      waylandSupport = true;
    };

    prismlauncher-qt5 = prev.prismlauncher-qt5.override {
      glfw = inputs.nixpkgs-glfw.legacyPackages.${prev.system}.glfw;
    };\
```
I can actually get the game to run through XWayland but the flickering isn't as intense as it was when I just use either the library glfw or the current nixpkg glfw.

**Comment by getchoo** [comment]
could you run from the branch directly with this expression (and `nix run -f <file>.nix`)? this is what works locally for me (note: this should only require building glfw, not all of prism)

```
let
  nixpkgs = builtins.fetchTarball {
    url = "https://github.com/getchoo/nixpkgs/archive/59f29ad755ff471b94ab65ec433f4665c120c41b.tar.gz";
    sha256 = "19p0y0ib7ds4ligrrsaf32yv55zgqjz36s0kidm2iyk8i074ivhm";
  };

  overlay = final: prev: {
    prismlauncher = prev.prismlauncher.override { glfw = final.glfw-wayland; };
  };
in
  (import nixpkgs {
    system = builtins.currentSystem;
    config = { };
    overlays = [ overlay ];
  })
  .prismlauncher
```

this could also be a difference in desktop and graphics driver, though. it works great with gnome and plasma 6 on nvidia 550.54.14 - which is probably one of the larger edge cases - so i'd be surprised if it's broken anywhere else

**Comment by NovaViper** [comment]
> it works great with gnome and plasma 6 on nvidia 550.54.14 - which is probably one of the larger edge cases - so i'd be surprised if it's broken anywhere else

Extremely weird... I'm running Plasma 6 Wayland with a RTX 2060 (Vulkan_beta build v550.40.55) and it flickers to death for me. Maybe it could be because I'm using the vulkan_beta drivers and not the latest or the production drivers?

> could you run from the branch directly with this expression (and nix run -f <file>.nix)? this is what works locally for me (note: this should only require building glfw, not all of prism)

Gave it a try but it still crashes for me with that window icon error still.. I'm gonna go and switch my nvidia drviers just in case

**Comment by NovaViper** [comment]
Just gave that expression a test with the latest drivers (with `package = config.boot.kernelPackages.nvidiaPackages.latest` and not `package = config.boot.kernelPackages.nvidiaPackages.vulkan_beta``) but it sadly still crashes with that same error `GLFW error 65548: Wayland: The platform does not support setting the window icon`
PrismLauncher config showing it's using the PR glfw version
image

Quick question, what specific NixOS unstable commit you're using specifically? I wonder if that could be affecting it (since I'm using `nixos-unstable-small` nixpkgs and it's currently at `24.05.20240315.2bcf18c`

**Comment by getchoo** [comment]
> and it flickers to death for me

this makes me think you're still on xwayland, as that's a pretty well known issue there. could you confirm with the kwin debug console -- just as a sanity check?

> Quick question, what specific NixOS unstable commit you're using specifically?

d691274a972b3165335d261cc4671335f5c67de9 on nixos-unstable

**Comment by NovaViper** [comment]
> this makes me think you're still on xwayland, as that's a pretty well known issue there. could you confirm with the kwin debug console -- just as a sanity check?

Yeah the flickering occurs when I run Minecraft through XWayland specifically. But when I use the PR glfw and launch it with Wayland support, it just crashes with that window icon error. But when I use that `glfw-wayland-minecraft` package, it doesn't crash (as it launches in Wayland) and doesn't flicker at all and runs fine (except for that "cannot set cursor position" error that appears every time I interact with the menu while the game's running.

**Comment by fpletz** [comment]
> I would also suggest following upstream here and enabling both X11 and Wayland on supported platforms (which is done by default), removing the need for the `glfw-wayland` package; this hasn't been implemented for now though as I wanted to get some feedback.

:+1: IMHO we should follow upstream here. It makes things so much easier for wayland users.

File: pkgs/development/libraries/glfw/3.x.nix [hunk 1]
```diff
@@ -29,16 +29,18 @@ stdenv.mkDerivation rec {
 
   buildInputs =
     lib.optionals stdenv.isDarwin [ Carbon Cocoa Kernel ]
-    ++ lib.optionals (stdenv.isLinux && waylandSupport) [ wayland wayland-protocols libxkbcommon ]
+    ++ lib.optionals (stdenv.isLinux && waylandSupport) [ wayland wayland-scanner wayland-protocols libxkbcommon ]
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/glfw/3.x.nix]: wayland-scanner belongs into nativeBuildInputs

File: pkgs/top-level/aliases.nix [hunk 2]
```diff
@@ -355,6 +355,7 @@ mapAliases ({
 
   gitter = throw "gitter has been removed since the client has been abandoned by upstream with the backend migration to Matrix"; # Added 2023-09-18
   glide = throw "'glide' has been removed as it is unmaintained, please use Go modules instead"; # Added 2023-12-26
+  glfw-wayland = throw "'glfw-wayland' has been removed as upstream now builds with Wayland by default. It is replaced by the regular 'glfw' package"; # Added 2024-03-31
```
Comment (SuperSandro2000) [inline pkgs/top-level/aliases.nix:358]: We can just alias this

File: pkgs/top-level/aliases.nix [hunk 3]
```diff
@@ -1168,6 +1168,7 @@ mapAliases ({
   virtmanager-qt = throw "'virtmanager-qt' has been renamed to/replaced by 'virt-manager-qt'"; # Converted to throw 2023-09-10
   vivaldi-widevine = throw "'vivaldi-widevine' has been renamed to/replaced by 'widevine-cdm'"; # Added 2023-02-25
   vkBasalt = vkbasalt; # Added 2022-11-22
+  vkdt-wayland = throw "'vkdt-wayland' has been replaced by 'vkdt'"; # Added 2024-03-31
```
Comment (SuperSandro2000) [inline pkgs/top-level/aliases.nix:1171]: we can just alias this, too.

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by oxalica** [comment]
> (but the `65548: Wayland: The platform does not support setting the cursor position` error appears in the logs whenever I open any of the GUI elements like the inventory). Looks like the minecraft version of `glfw-wayland` that's currently in nixpkgs is actually old and [hasn't had the patch needed to make the cursor position work properly](https://github.com/Admicos/minecraft-wayland/pull/56).

To anyone who may concern, I opened #306264 to update `glfw-wayland-minecraft` to glfw 3.4 with that set-cursor-position patch applied.

**Comment by roman3pm** [comment]
Hi, I have a little project on OpenGL. I used `glfw-wayland` and it worked just fine. But after the GLFW 3.4 update, it refuses to start with a GLFW initialization error. The same error occurred if I tried to use `glfw` (without `-wayland`) from the stable branch. And it worked if I used `glfw-wayland-minecraft`. I have to say that something definitely went wrong with GLFW 3.4 for Wayland users.

**Pull request merged**